testing/internal/e2e/tests/static/target_address_test.go

```diff
@@ -0,0 +1,342 @@
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+	newHostCatalogId := boundary.CreateNewHostCatalogCli(t, ctx, newProjectId)
+	newHostSetId := boundary.CreateNewHostSetCli(t, ctx, newHostCatalogId)
+	newHostId := boundary.CreateNewHostCli(t, ctx, newHostCatalogId, c.TargetIp)
+	boundary.AddHostToHostSetCli(t, ctx, newHostSetId, newHostId)
+	newTargetId := boundary.CreateNewAddressTargetCli(t, ctx, newProjectId, c.TargetPort, c.TargetIp)
+
+	// Connect to target and print host's IP address
+	output := e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"connect",
+			"-target-id", newTargetId,
+			"-exec", "/usr/bin/ssh", "--",
+			"-l", c.TargetSshUser,
+			"-i", c.TargetSshKeyPath,
+			"-o", "UserKnownHostsFile=/dev/null",
+			"-o", "StrictHostKeyChecking=no",
+			"-o", "IdentitiesOnly=yes", // forces the use of the provided key
+			"-p", "{{boundary.port}}", // this is provided by boundary
+			"{{boundary.ip}}",
+			"hostname", "-i",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	parts := strings.Fields(string(output.Stdout))
+	hostIp := parts[len(parts)-1]
+	require.Equal(t, c.TargetIp, hostIp, "SSH session did not return expected output")
+	t.Log("Successfully connected to target")
+
+	// Attempt add-host-sources, which appends a new host source to an existing host sources.
+	// should error because the target has an address.
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "add-host-sources",
+			"-id", newTargetId,
+			"-host-source", newHostSetId,
+			"-format", "json",
+		),
+	)
+	require.Error(t, output.Err, "Target add-host-sources did not return expected error")
+
+	var response boundary.CliError
+	require.NoError(t, json.Unmarshal(output.Stderr, &response))
+	require.Equal(t, http.StatusBadRequest, response.Status, "Expected error when adding host source to target with an address")
+
+	// Attempt set-host-sources, which replaces all existing host source(s) with a new one.
+	// should error because the target has an address.
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "set-host-sources",
+			"-id", newTargetId,
+			"-host-source", newHostSetId,
+			"-format", "json",
+		),
+	)
+	require.Error(t, output.Err, "Target set-host-sources did not return expected error")
+	require.NoError(t, json.Unmarshal(output.Stderr, &response))
+	require.Equal(t, http.StatusBadRequest, response.Status, "Expected error when setting host source to target with an address")
+
+	// Remove the target's address. We should now be able to add/set host
+	// sources.
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "update", "tcp",
+			"-id", newTargetId,
+			"-address", "null",
+			"-format", "json",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	boundary.AddHostSourceToTargetCli(t, ctx, newTargetId, newHostSetId)
+	boundary.SetHostSourceToTargetCli(t, ctx, newTargetId, newHostSetId)
```

Comment:
Just curious, why do we do both `add` and `set` here? You just want to confirm that they perform successfully?

Reply:
@moduli Correct. `add` and `set` are two entirely different Boundary operations that have their own codepaths, and we added stuff on both, so we have to test both.

Reply:
Do we need to confirm that it was `added` correctly? (outside of just verifying that the command didn't error, that the `add` operation correctly added the host source.

Reply:
That's we try a `connect` afterwards. If it wasn't added or set correctly, the connect won't find any host source to connect to 🙂

Reply:
Ah! I see what you mean now. Yeah, the set will replace what the add did, so here we're relying on an error to be given back from the CLI call. We can always run 2 boundary connects, one after adding, then again after setting 😄